qiskit/quantum_info/synthesis/two_qubit_decompose.py

```diff
@@ -926,9 +949,257 @@ def __call__(self, target, basis_fidelity=None, *, _num_basis_uses=None) -> Quan
+            )
+        if self._decomposer1q.basis in {"ZSX", "ZSXX"}:
+            if isinstance(self.gate, CXGate):
+                if best_nbasis == 3:
+                    circuit = self._get_sx_vz_3cx_efficient_euler(decomposition, target_decomposed)
+                elif best_nbasis == 2:
+                    circuit = self._get_sx_vz_2cx_efficient_euler(decomposition, target_decomposed)
+            else:
+                raise QiskitError("pulse_optimizer currently only works with CNOT entangling gate")
+        else:
+            raise QiskitError(
+                '"pulse_optimize" currently only works with ZSX basis '
+                f"({self._decomposer1q.basis} used)"
+            )
+        return circuit
+
+    def _get_sx_vz_2cx_efficient_euler(self, decomposition, target_decomposed):
+        """
+        Decomposition of SU(4) gate for device with SX, virtual RZ, and CNOT gates assuming
+        two CNOT gates are needed.
+
+        This first decomposes each unitary from the KAK decomposition into ZXZ on the source
+        qubit of the CNOTs and XZX on the targets in order to commute operators to beginning and
+        end of decomposition. The beginning and ending single qubit gates are then
+        collapsed and re-decomposed with the single qubit decomposer. This last step could be avoided
+        if performance is a concern.
+        """
+        best_nbasis = 2  # by assumption
+        num_1q_uni = len(decomposition)
+        # list of euler angle decompositions on qubits 0 and 1
+        euler_q0 = np.empty((num_1q_uni // 2, 3), dtype=float)
+        euler_q1 = np.empty((num_1q_uni // 2, 3), dtype=float)
+        global_phase = 0.0
+
+        # decompose source unitaries to zxz
+        zxz_decomposer = OneQubitEulerDecomposer("ZXZ")
+        for iqubit, decomp in enumerate(decomposition[0::2]):
+            euler_angles = zxz_decomposer.angles_and_phase(decomp)
+            euler_q0[iqubit, [1, 2, 0]] = euler_angles[:3]
+            global_phase += euler_angles[3]
+        # decompose target unitaries to xzx
+        xzx_decomposer = OneQubitEulerDecomposer("XZX")
+        for iqubit, decomp in enumerate(decomposition[1::2]):
+            euler_angles = xzx_decomposer.angles_and_phase(decomp)
+            euler_q1[iqubit, [1, 2, 0]] = euler_angles[:3]
+            global_phase += euler_angles[3]
+        qc = QuantumCircuit(2)
+        qc.global_phase = target_decomposed.global_phase
+        qc.global_phase -= best_nbasis * self.basis.global_phase
+        qc.global_phase += global_phase
+
+        # TODO: make this more effecient to avoid double decomposition
+        # prepare beginning 0th qubit local unitary
+        circ = QuantumCircuit(1)
+        circ.rz(euler_q0[0][0], 0)
+        circ.rx(euler_q0[0][1], 0)
+        circ.rz(euler_q0[0][2] + euler_q0[1][0] + math.pi / 2, 0)
+        # re-decompose to basis of 1q decomposer
+        qceuler = self._decomposer1q(Operator(circ).data)
```

Comment:
This is just a note (probably for myself) for the the future performance tuning on this. This is fine as is.

For performance it would probably be faster to do:

```suggestion
        circ = RZGate(euler_q0[0][0]).to_matrix()
        circ.dot(RXGate(euler_q0[0][1]).to_matrix())
        circ.dot(RZGate(euler_q0[0][2] + euler_q0[1][0] + math.pi / 2).to_matrix())
        # re-decompose to basis of 1q decomposer
        qceuler = self._decomposer1q(unitary)
```

Or even better would be to just inline the `__array__` from each gate class and construct the matrix manually without the overhead of object creation.

Reply:
The whole matrix multiplication and decomposition can be avoided by using the identity Rx(theta) = Rz(-pi/2).√X.Rz(pi-theta).√X.Rz(-pi/2) although one might have to manually catch some reductions for certain values of theta e.g. multiples of pi/2